### PR TITLE
DEV: Modernize the post menu from widgets to Glimmer components

### PR DESCRIPTION
**Pull request by megothss**

### CODE REVIEW

File: app/assets/javascripts/discourse/app/components/post/menu/buttons/bookmark.gjs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import Component from "@glimmer/component";
+import { getOwner } from "@ember/owner";
+import { inject as service } from "@ember/service";
+import BookmarkMenu from "discourse/components/bookmark-menu";
+import PostBookmarkManager from "../../../../lib/post-bookmark-manager";
```
Comment (megothss) [inline app/assets/javascripts/discourse/app/components/post/menu/buttons/bookmark.gjs:5]: Need to change this import to `discourse/lib/...`

File: app/assets/javascripts/discourse/app/widgets/post-menu.js [hunk 2]
```diff
@@ -949,3 +951,35 @@ export default createWidget("post-menu", {
...
+
+registerWidgetShim(
+  "glimmer-post-menu",
+  "section.post-menu-area.glimmer-post-menu-widget-shim.clearfix",
+  hbs`
+    <Post::Menu
+      @filteredRepliesView={{@data.filteredRepliesView}}
+      @model={{@data.model}}
+      @repliesShown={{@data.repliesShown}}
+      @transformedPost={{@data.transformedPost}}
```
Comment (davidtaylorhq) [inline app/assets/javascripts/discourse/app/widgets/post-menu.js:963]: `transformedPost` is something which was introduced for widgets. So ideally, we would avoid using it in any new code. Can we get access to the actual post model, and use that? @tyb-talks made a similar change for the glimmer-ified topic map [here](https://github.com/discourse/discourse/commit/edbd44e737753eb4e9bd91c7da1836954e0627db)

Comment (megothss) [reply app/assets/javascripts/discourse/app/widgets/post-menu.js]: I've removed `trasnformedPost` from the new components in [fe7acac](https://github.com/discourse/discourse/pull/28670/commits/fe7acac052aa7d3f66350076e1732b6df00eb6a2)

File: app/assets/javascripts/discourse/app/components/post/menu/buttons/bookmark.gjs (hunk 1 above)
Comment (megothss) [reply app/assets/javascripts/discourse/app/components/post/menu/buttons/bookmark.gjs]: Done

File: app/assets/javascripts/discourse/app/components/admin-post-menu.gjs [hunk 3]
```diff
@@ -14,13 +14,13 @@ export default class AdminPostMenu extends Component {
   @service adminPostMenuButtons;
 
   get reviewUrl() {
-    return `/review?topic_id=${this.args.data.transformedPost.id}&status=all`;
+    return `/review?topic_id=${this.args.data.id}&status=all`;
```
Comment (davidtaylorhq) [inline app/assets/javascripts/discourse/app/components/admin-post-menu.gjs:17]:
```suggestion
    return `/review?topic_id=${this.args.data.post.id}&status=all`;
```
? (to match all the other changes in this file)

File: app/assets/javascripts/discourse/app/components/admin-post-menu.gjs
Comment (davidtaylorhq) [inline app/assets/javascripts/discourse/app/components/admin-post-menu.gjs:1]: Might be nice to land these AdminPostMenu changes in a separate PR? They're good, but not strictly related to the post menu change.

File: app/assets/javascripts/discourse/app/components/post/menu.gjs [hunk 4]
```diff
@@ -0,0 +1,675 @@
...
+    if (!this.collapsed) {
+      await this.#fetchWhoLiked();
+    }
+  }
+
+  get #configuredItems() {
+    return this.siteSettings.post_menu.split("|").filter(Boolean);
+  }
+
+  @computed("args.post.bookmarked")
```
Comment (davidtaylorhq) [inline app/assets/javascripts/discourse/app/components/post/menu.gjs:554]: We should avoid using legacy `@computed` in new code.

If you do `this.args.post.get("bookmarked")`, then it should autotrack properly

(or even better, add `@tracked bookmarked` to the post model)

File: app/assets/javascripts/discourse/app/components/post/menu/buttons/like.gjs [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+  @bind
+  setElement(element) {
+    this.#element = element;
+  }
+
+  @action
+  animateToggle() {
+    this.#element.classList.add("has-like");
+    this.#element.querySelector(`.d-icon`).classList.add("heart-animation");
+  }
```
Comment (davidtaylorhq) [inline app/assets/javascripts/discourse/app/components/post/menu/buttons/like.gjs:50]: Storing a reference and then manipulating later is a bit of a code smell.

I think a more idiomatic way would be to do something like

```js
@tracked animating;

@action
animateToggle(){
  this.animating = true;
}
```

and then put the class logic in the hbs like

```hbs
class={{if this.animating "has-like heart-animation"}}
```

File: app/assets/javascripts/discourse/app/components/admin-post-menu.gjs
Comment (megothss) [reply app/assets/javascripts/discourse/app/components/admin-post-menu.gjs:1]: Oh, but they are. 

I changed it because the Glimmer component calls de AdminPostMenu and doesn't have the `transformedPost` post to pass around.

File: app/assets/javascripts/discourse/app/components/post/menu.gjs (hunk 4 above)
Comment (megothss) [reply app/assets/javascripts/discourse/app/components/post/menu.gjs:424]: Done

File: app/assets/javascripts/discourse/app/components/post/menu/buttons/like.gjs (hunk 5 above)
Comment (megothss) [reply app/assets/javascripts/discourse/app/components/post/menu/buttons/like.gjs]: Done

File: app/assets/javascripts/discourse/app/components/admin-post-menu.gjs
Comment (davidtaylorhq) [reply app/assets/javascripts/discourse/app/components/admin-post-menu.gjs:1]: Ahhh ok, sounds good

File: app/assets/javascripts/discourse/app/components/post/menu/button-wrapper.gjs [hunk 6]
```diff
@@ -0,0 +1,19 @@
+import Component from "@glimmer/component";
+
+// eslint-disable-next-line ember/no-empty-glimmer-component-classes
+export default class PostMenuButtonWrapper extends Component {
+  // we need a class component because we need to pass this.args to the config helpers
+
+  <template>
+    <@buttonConfig.Component
+      class="btn-flat"
+      ...attributes
```
Comment (davidtaylorhq) [inline app/assets/javascripts/discourse/app/components/post/menu/button-wrapper.gjs:10]: I don't see anything actually passing attributes in here? IMO it's better to drop the `...attributes` in this case - then it's clear that we don't expect anything to be passed in.

File: app/assets/javascripts/discourse/app/components/post/menu/buttons/admin.gjs [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    return (
+      args.post.canManage || args.post.can_wiki || args.post.canEditStaffNotes
+    );
+  }
+
+  <template>
+    {{#if @shouldRender}}
+      <DButton
+        class="post-action-menu__admin show-post-admin-menu"
+        ...attributes
```
Comment (davidtaylorhq) [inline app/assets/javascripts/discourse/app/components/post/menu/buttons/admin.gjs:15]: Similar to the previous comment - I don't think this is actually being used? Can we drop it in all the buttons for simplicity/readability?

File: app/assets/javascripts/discourse/app/lib/transform-post.js [hunk 8]
```diff
@@ -53,7 +53,7 @@ export function transformBasicPost(post) {
     title_is_group: post.title_is_group,
     created_at: post.created_at,
     updated_at: post.updated_at,
-    canDelete: post.can_delete,
+    // canDelete: post.can_delete,
```
Comment (davidtaylorhq) [inline app/assets/javascripts/discourse/app/lib/transform-post.js:56]: ?

File: app/assets/javascripts/discourse/app/components/post/menu/button-wrapper.gjs [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+      @alwaysShow={{@buttonConfig.alwaysShow this.args}}
+      @buttonActions={{@buttonActions}}
+      @context={{@context}}
+      @post={{@post}}
+      @shouldRender={{@buttonConfig.shouldRender this.args}}
```
Comment (davidtaylorhq) [inline app/assets/javascripts/discourse/app/components/post/menu/button-wrapper.gjs:15]: The shouldRender architecture feels a little weird to me. Buttons define a static `shouldRender` method, which is evaluated outside the component, then passed back into the button component, which then has to wrap itself in a `{{#if shouldRender}}`

IMO it would be much cleaner if the `{{#if shouldRender}}` check was done in this ButtonWrapper component, so that the button component is not rendered when shouldRender is false.

IIRC the current architecture is so that themes/plugins can replace the component, and choose to ignore the `shouldRender` value? Maybe we could make a way for them to override the shouldRender function instead?

Comment (davidtaylorhq) [reply app/assets/javascripts/discourse/app/components/post/menu/button-wrapper.gjs:15]: Also, rendering every single button component even when it's not required is quite a large cost. If we can apply the shouldRender logic earlier, it should make things faster.

Comment (megothss) [reply app/assets/javascripts/discourse/app/components/post/menu/button-wrapper.gjs:15]: Your point about the `shouldRender` logic being too convoluted and the performance issues, makes total sense.

I'll think of a way to prevent the buttons from being rendered, and if the user chooses to override the shouldRender static method, we still be able to provide the core `shouldRender` value to the component.

That would allow it to disabled it instead of hiding for example

File: app/assets/javascripts/discourse/app/components/post/menu/button-wrapper.gjs (hunk 6 above)
Comment (megothss) [reply app/assets/javascripts/discourse/app/components/post/menu/button-wrapper.gjs:10]: We can drop `...attributes` here as I probably just left it as default because I find annoying when I have to open a PR just to add `...attributes` to a component in order to customize it.

But you're right, it's not needed here.

File: app/assets/javascripts/discourse/app/components/post/menu/buttons/admin.gjs (hunk 7 above)
Comment (megothss) [reply app/assets/javascripts/discourse/app/components/post/menu/buttons/admin.gjs:15]: We can't remove it because on the buttons `...attributes` is being used to propagate the `btn-flat` class that is applied in the `PostMenuButtonWrapper` component

File: app/assets/javascripts/discourse/app/lib/transformer/registry.js [hunk 10]
```diff
@@ -11,4 +11,5 @@ export const VALUE_TRANSFORMERS = Object.freeze([
   "header-notifications-avatar-size",
   "home-logo-href",
   "home-logo-image-url",
+  "post-menu-registered-buttons",
```
[the remark this answers is not in the document]
Comment (megothss) [reply app/assets/javascripts/discourse/app/lib/transformer/registry.js]: Yep. I changed it.

File: app/assets/javascripts/discourse/app/lib/transform-post.js (hunk 8 above)
Comment (megothss) [reply app/assets/javascripts/discourse/app/lib/transform-post.js]: Probably just a left over from some experiment.

File: app/assets/javascripts/discourse/app/components/post/menu/buttons/bookmark.gjs [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+    // lazy instantiate the bookmark manager only if the component is rendered
+    if (!this.#bookmarkManager) {
+      this.#bookmarkManager = new PostBookmarkManager(
+        getOwner(this),
+        this.args.post
+      );
+    }
+
+    return this.#bookmarkManager;
+  }
```
[the remark this answers is not in the document]
Comment (megothss) [reply app/assets/javascripts/discourse/app/components/post/menu/buttons/bookmark.gjs]: Nice catch! I changed it.

File: app/assets/javascripts/discourse/app/components/post/menu/button-wrapper.gjs (hunk 9 above)
Comment (megothss) [reply app/assets/javascripts/discourse/app/components/post/menu/button-wrapper.gjs]: I introduced a new static property in the component called `delegateShouldRenderToTemplate`. When it's set to `true` the component is rendered and the value of `shouldRender` is passed as an argument.

This solves both issues, we no longer render the buttons and check if it should be rendered in the template, and as consequence, we can remove the  `{{#if @shouldRender}}` from the core buttons

File: app/assets/javascripts/discourse/app/components/post/menu/button-wrapper.gjs (hunk 6 above)
Comment (megothss) [reply app/assets/javascripts/discourse/app/components/post/menu/button-wrapper.gjs]: Done

File: app/assets/javascripts/discourse/app/components/post/menu.gjs [hunk 12]
```diff
@@ -0,0 +1,557 @@
...
+
+  @tracked collapsed = true; // TODO some plugins will need a value transformer
+  @tracked isWhoLikedVisible = false;
+  @tracked likedUsers = [];
+  @tracked totalLikedUsers;
+  @tracked isWhoReadVisible = false;
+  @tracked readers = [];
+  @tracked totalReaders;
+
+  get buttonActions() {
```
Comment (davidtaylorhq) [inline app/assets/javascripts/discourse/app/components/post/menu.gjs:93]: We pass this to every button, so I think `@cached` would be good for perf

File: app/assets/javascripts/discourse/app/components/post/menu.gjs [hunk 13]
```diff
@@ -0,0 +1,557 @@
...
+        this.args.nextPost?.reply_to_post_number ===
+          this.args.post.post_number &&
+        this.args.post.post_number !== this.args.post.filteredRepliesPostNumber,
+      showReadIndicator: this.args.showReadIndicator,
+      suppressReplyDirectlyBelow:
+        this.siteSettings.suppress_reply_directly_below,
+    });
+  }
+
+  get staticMethodsArgs() {
```
Comment (davidtaylorhq) [inline app/assets/javascripts/discourse/app/components/post/menu.gjs:134]:
```suggestion
  @cached
  get staticMethodsArgs() {
```

`@cached` for perf?

File: app/assets/javascripts/discourse/app/components/post/menu.gjs [hunk 14]
```diff
@@ -0,0 +1,557 @@
...
+    return Object.freeze({
+      context: this.staticMethodsContext,
+      post: this.args.post,
+    });
+  }
+
+  get context() {
```
Comment (davidtaylorhq) [inline app/assets/javascripts/discourse/app/components/post/menu.gjs:141]:
```suggestion
  @cached
  get context() {
```

File: app/assets/javascripts/discourse/app/components/post/menu.gjs [hunk 15]
```diff
@@ -0,0 +1,557 @@
...
+    });
+  }
+
+  @cached
+  get registeredButtons() {
+    // the DAG is not resolved now, instead we just use the object for convenience to pass a nice DAG API to be used
+    // in the value transformer, and extract the data to be used later to resolve the DAG order
+    const buttonsRegistry = applyValueTransformer(
+      "post-menu-buttons",
+      // TODO auto-generate the position coordinates based on the order of the elements
```
Comment (davidtaylorhq) [inline app/assets/javascripts/discourse/app/components/post/menu.gjs:154]: Do we need to resolve this before merging?

File: app/assets/javascripts/discourse/app/widgets/actions-summary.js [hunk 16]
```diff
@@ -1,40 +1,24 @@
...
-        {
-          attrs,
-        }
-      ),
-    ];
-  },
-});
+// TODO check the rendering in the two voting plugins
+// This shim is nesting everything into a DIV and changing the HTML but only thw two voting plugins
+// are using this widget outside of core.
```
Comment (davidtaylorhq) [inline app/assets/javascripts/discourse/app/widgets/actions-summary.js:6]: Should we check the voting plugins before merging?

File: app/assets/javascripts/discourse/app/widgets/post-user-tip-shim.js [hunk 17]
```diff
@@ -1,6 +1,8 @@
 import { hbs } from "ember-cli-htmlbars";
 import { registerWidgetShim } from "discourse/widgets/render-glimmer";
 
+// TODO deprecate for the Glimmer Post Menu
```
Comment (davidtaylorhq) [inline app/assets/javascripts/discourse/app/widgets/post-user-tip-shim.js:4]: What does this mean? We need to add it to the list of deprecated widgets in plugin-api.js?

File: app/assets/javascripts/discourse/app/components/post/menu.gjs [hunk 18]
```diff
@@ -0,0 +1,557 @@
...
+            "showLabel",
+          ].forEach((method) => {
+            if (
+              typeof newComponent[method] === "undefined" &&
+              typeof oldComponent[method] !== "undefined"
+            ) {
+              newComponent[method] = oldComponent[method];
+            }
+          });
+        },
```
[the remark this answers is not in the document]
Comment (megothss) [reply app/assets/javascripts/discourse/app/components/post/menu.gjs]: I changed the button config to use a weak map that stores references to the old button class instead.

File: app/assets/javascripts/discourse/app/components/post/menu.gjs (hunk 15 above)
Comment (megothss) [reply app/assets/javascripts/discourse/app/components/post/menu.gjs]: It's resolved. The positions are now auto-generated based on the configured items.

File: app/assets/javascripts/discourse/app/widgets/actions-summary.js (hunk 16 above)
Comment (megothss) [reply app/assets/javascripts/discourse/app/widgets/actions-summary.js:6]: I checked how both plugins are rendering the widget and it's fine.

File: app/assets/javascripts/discourse/app/widgets/post-user-tip-shim.js (hunk 17 above)
Comment (megothss) [reply app/assets/javascripts/discourse/app/widgets/post-user-tip-shim.js:4]: Exactly. I added the widget to the deprecated list

File: app/assets/javascripts/discourse/app/components/post/menu.gjs [hunk 19]
```diff
@@ -498,6 +499,10 @@ export default class PostMenu extends Component {
   <template>
     {{! The section tag can't be include while we're still using the widget shim }}
     {{! <section class="post-menu-area clearfix"> }}
+    <PluginOutlet
+      @name="post-menu__before"
+      @outletArgs={{hash context=this.context post=@post}}
+    />
```
Comment (davidtaylorhq) [inline app/assets/javascripts/discourse/app/components/post/menu.gjs:505]: Should we make this a wrapper plugin outlet? Then we get the `__before` and `__after` for free?

Comment (megothss) [reply app/assets/javascripts/discourse/app/components/post/menu.gjs]: Yep. I replaced them

File: app/assets/javascripts/discourse/app/components/post/menu.gjs [hunk 20]
```diff
@@ -0,0 +1,603 @@
...
+    >
+      <nav
+        class={{concatClass
+          "post-controls"
+          "glimmer-post-menu"
+          (if
+            (and
+              (this.showMoreButton.shouldRender
+                (hash post=this.post state=this.state)
+              )
```
Comment (CvX) [inline app/assets/javascripts/discourse/app/components/post/menu.gjs:519]: could you extract it to a getter? calling `shouldRender` like a helper means it won't have the correct `this` context, so you won't be able to e.g. call other static methods on a given class

Comment (megothss) [reply app/assets/javascripts/discourse/app/components/post/menu.gjs:519]: This is not an issue, `this.showMoreButton` in this case is a `PostMenuButtonConfig` instance.

If we were indeed calling the static method we'd also have to pass the other parameters: `(args, context, owner)`. :-)

The static method `shouldRender` is called internally by the button config `shouldRender` method.

https://github.com/discourse/discourse/blob/6e7f27648fa58b1c98ad08dddadd701fdf75c433/app/assets/javascripts/discourse/app/components/post/menu/button-config.gjs#L64-L71

**Review by davidtaylorhq — APPROVED** [review]

File: app/assets/javascripts/discourse/app/lib/plugin-api.gjs [hunk 21]
```diff
@@ -3,7 +3,7 @@
 // docs/CHANGELOG-JAVASCRIPT-PLUGIN-API.md whenever you change the version
 // using the format described at https://keepachangelog.com/en/1.0.0/.
 
-export const PLUGIN_API_VERSION = "1.38.0";
+export const PLUGIN_API_VERSION = "1.37.4";
```
Comment (davidtaylorhq) [inline app/assets/javascripts/discourse/app/lib/plugin-api.gjs:6]: Looks like core got bumped, so we'll need to bump it higher in this branch

File: app/assets/javascripts/discourse/app/widgets/post.js [hunk 22]
```diff
@@ -41,6 +43,19 @@ function transformWithCallbacks(post, topicUrl, store) {
   return transformed;
 }
 
+let postMenuWidgetExtensionsAdded = null;
+let postMenuConsoleWarningLogged = false;
+
+registerDeprecationHandler((_, opts) => {
```
Comment (davidtaylorhq) [inline app/assets/javascripts/discourse/app/widgets/post.js:49]: Since we added a global silence for the deprecation... I don't think this handler will ever fire?

I guess that's ok. It just means that the 'auto' mode won't work properly until we remove that global silence.

Should we unsilence globally before merging this PR? Or you prefer to do it later?

**Review by davidtaylorhq — APPROVED** [review]

File: app/assets/javascripts/discourse/app/lib/plugin-api.gjs (hunk 21 above)
Comment (megothss) [reply app/assets/javascripts/discourse/app/lib/plugin-api.gjs]: Done
